File: backend/src/usecase/auth_service.rs
use std::sync::Arc;

use uuid::Uuid;

use crate::domain::entity::refresh_token::RefreshToken;
use crate::domain::valueobject::email::Email;
use crate::domain::valueobject::jwt::{JwtError, JwtService, TokenPair};
use crate::repository::refresh_token_repository::{
    RefreshTokenRepository, RefreshTokenRepositoryError,
};
use crate::repository::user_repository::{UserRepository, UserRepositoryError};

/// 認証エラー
#[derive(Debug, thiserror::Error)]
pub enum AuthError {
    #[error("Invalid credentials")]
    InvalidCredentials,
    #[error("Invalid token")]
    InvalidToken,
    #[error("Token expired")]
    TokenExpired,
    #[error("Token revoked")]
    TokenRevoked,
    #[error("User not found")]
    UserNotFound,
    #[error("Repository error: {0}")]
    Repository(String),
    #[error("JWT error: {0}")]
    Jwt(String),
}

impl From<UserRepositoryError> for AuthError {
    fn from(err: UserRepositoryError) -> Self {
        Self::Repository(err.to_string())
    }
}

impl From<RefreshTokenRepositoryError> for AuthError {
    fn from(err: RefreshTokenRepositoryError) -> Self {
        Self::Repository(err.to_string())
    }
}

impl From<JwtError> for AuthError {
    fn from(err: JwtError) -> Self {
        match err {
            JwtError::Expired => Self::TokenExpired,
            JwtError::InvalidTokenType => Self::InvalidToken,
            _ => Self::Jwt(err.to_string()),
        }
    }
}

/// ログインリクエスト
#[derive(Debug, Clone)]
pub struct LoginRequest {
    pub email: String,
    pub password: String,
}

/// ログインレスポンス
#[derive(Debug, Clone)]
pub struct LoginResponse {
    pub user_id: String,
    pub username: String,
    pub email: String,
    pub token_pair: TokenPair,
}

/// 認証サービス
pub struct AuthService<U: UserRepository, R: RefreshTokenRepository> {
    user_repository: Arc<U>,
    refresh_token_repository: Arc<R>,
    jwt_service: JwtService,
}

impl<U: UserRepository, R: RefreshTokenRepository> AuthService<U, R> {
    /// 新しい認証サービスを作成
    #[must_use]
    pub fn new(
        user_repository: Arc<U>,
        refresh_token_repository: Arc<R>,
        jwt_secret: String,
    ) -> Self {
        Self {
            user_repository,
            refresh_token_repository,
            jwt_service: JwtService::new(jwt_secret),
        }
    }

    /// ログイン処理：メール+パスワード → TokenPair + DB保存
    ///
    /// # Errors
    ///
    /// - 認証情報が無効な場合
    /// - DB操作に失敗した場合
    pub async fn login(&self, request: LoginRequest) -> Result<LoginResponse, AuthError> {
        // 1. メールアドレスのバリデーション
        let email = Email::new(&request.email).map_err(|_| AuthError::InvalidCredentials)?;

        // 2. ユーザーの取得
        let user = self
            .user_repository
            .find_by_email(&email)
            .await?
            .ok_or(AuthError::InvalidCredentials)?;

        // 3. パスワードの検証
        let is_valid = user
            .verify_password(&request.password)
            .map_err(|_| AuthError::InvalidCredentials)?;
        if !is_valid {
            return Err(AuthError::InvalidCredentials);
        }

        // 4. Token Pairを生成
        let token_pair = self.jwt_service.generate_token_pair(user.user_id())?;

        // 5. Refresh TokenをDBに保存
        let token_hash = JwtService::hash_refresh_token(&token_pair.refresh_token);
        let refresh_token = RefreshToken::new(*user.user_id(), token_hash);
        self.refresh_token_repository.save(&refresh_token).await?;

        // 6. レスポンスを返す
        Ok(LoginResponse {
            user_id: user.user_id().to_string(),
            username: user.username().as_str().to_string(),
            email: user.email().as_str().to_string(),
            token_pair,
        })
    }

    /// トークンリフレッシュ：古いRefresh Token → 新しいTokenPair
    ///
    /// # Errors
    ///
    /// - Refresh Tokenが無効な場合
    /// - Refresh TokenがDBに存在しない、または無効化されている場合
    pub async fn refresh(&self, refresh_token: &str) -> Result<TokenPair, AuthError> {
        // 1. Refresh Tokenを検証
        let claims = self.jwt_service.verify_refresh_token(refresh_token)?;
        let user_id =
            Uuid::parse_str(&claims.sub).map_err(|_| AuthError::InvalidToken)?;

        // 2. DBでトークンを確認
        let token_hash = JwtService::hash_refresh_token(refresh_token);
        let stored_token = self
            .refresh_token_repository
            .find_by_hash(&token_hash)
            .await?
            .ok_or(AuthError::InvalidToken)?;

        // 3. トークンが有効か確認
        if !stored_token.is_valid() {
            return Err(AuthError::TokenRevoked);
        }

        // 4. 古いトークンを削除（Token Rotation）
        self.refresh_token_repository
            .delete_by_id(stored_token.token_id())
            .await?;

        // 5. 新しいToken Pairを生成
        let new_token_pair = self.jwt_service.generate_token_pair(&user_id)?;

        // 6. 新しいRefresh TokenをDBに保存
        let new_token_hash = JwtService::hash_refresh_token(&new_token_pair.refresh_token);
        let new_refresh_token = RefreshToken::new(user_id, new_token_hash);
        self.refresh_token_repository.save(&new_refresh_token).await?;

        Ok(new_token_pair)
    }

    /// ログアウト：Refresh Tokenを無効化
    ///
    /// # Errors
    ///
    /// - DB操作に失敗した場合
    pub async fn logout(&self, refresh_token: &str) -> Result<(), AuthError> {
        let token_hash = JwtService::hash_refresh_token(refresh_token);

        if let Some(stored_token) = self
            .refresh_token_repository
            .find_by_hash(&token_hash)
            .await?
        {
            self.refresh_token_repository
                .delete_by_id(stored_token.token_id())
                .await?;
        }

        Ok(())
    }

    /// 全デバイスからログアウト
    ///
    /// # Errors
    ///
    /// - DB操作に失敗した場合
    pub async fn logout_all(&self, user_id: &Uuid) -> Result<(), AuthError> {
        self.refresh_token_repository
            .delete_all_by_user_id(user_id)
            .await?;
        Ok(())
    }

    /// Access Tokenを検証してユーザーIDを取得
    ///
    /// # Errors
    ///
    /// - トークンが無効な場合
    pub fn validate_access_token(&self, token: &str) -> Result<Uuid, AuthError> {
        let claims = self.jwt_service.verify_access_token(token)?;
        Uuid::parse_str(&claims.sub).map_err(|_| AuthError::InvalidToken)
    }
}

#[cfg(test)]
mod tests {
    // テストは後で追加
}

File: backend-hono/src/usecase/AuthService.ts
import { v4 as uuidv4 } from 'uuid';
import { User } from '../domain/entity/User';
import { RefreshToken } from '../domain/entity/RefreshToken';
import { UserRepository } from '../repository/interface/UserRepository';
import { RefreshTokenRepository } from '../repository/interface/RefreshTokenRepository';
import { JwtService, TokenPair } from '../domain/valueobject/JWT';

export class AuthError extends Error {
  constructor(
    message: string,
    public readonly code: string
  ) {
    super(message);
    this.name = 'AuthError';
  }

  static invalidCredentials(): AuthError {
    return new AuthError('Invalid email or password', 'INVALID_CREDENTIALS');
  }

  static invalidToken(): AuthError {
    return new AuthError('Invalid token', 'INVALID_TOKEN');
  }

  static tokenExpired(): AuthError {
    return new AuthError('Token expired', 'TOKEN_EXPIRED');
  }

  static tokenRevoked(): AuthError {
    return new AuthError('Token has been revoked', 'TOKEN_REVOKED');
  }

  static userNotFound(): AuthError {
    return new AuthError('User not found', 'USER_NOT_FOUND');
  }
}

export interface RegisterRequest {
  username: string;
  email: string;
  password: string;
}

export interface RegisterResponse {
  userId: string;
  username: string;
  email: string;
}

export interface LoginRequest {
  email: string;
  password: string;
}

export interface LoginResponse {
  userId: string;
  username: string;
  email: string;
  accessToken: string;
  refreshToken: string;
  tokenType: string;
  expiresIn: number;
}

export class AuthService {
  constructor(
    private readonly userRepository: UserRepository,
    private readonly refreshTokenRepository: RefreshTokenRepository,
    private readonly jwtService: JwtService
  ) {}

  async register(request: RegisterRequest): Promise<RegisterResponse> {
    // Check if username or email already exists
    const usernameExists = await this.userRepository.existsByUsername(request.username);
    if (usernameExists) {
      throw new AuthError('Username already exists', 'USERNAME_EXISTS');
    }

    const emailExists = await this.userRepository.existsByEmail(request.email);
    if (emailExists) {
      throw new AuthError('Email already exists', 'EMAIL_EXISTS');
    }

    // Create user
    const userId = uuidv4();
    const user = await User.create(userId, request.username, request.email, request.password);

    // Save user
    await this.userRepository.save(user);

    return {
      userId: user.getUserId(),
      username: user.getUsername(),
      email: user.getEmail(),
    };
  }

  async login(request: LoginRequest): Promise<LoginResponse> {
    // Find user by email
    const user = await this.userRepository.findByEmail(request.email);
    if (!user) {
      throw AuthError.invalidCredentials();
    }

    // Verify password
    const isValid = await user.verifyPassword(request.password);
    if (!isValid) {
      throw AuthError.invalidCredentials();
    }

    // Generate tokens
    const tokenPair = this.jwtService.generatePair(user.getUserId());

    // Save refresh token
    const tokenId = uuidv4();
    const refreshToken = RefreshToken.create(tokenId, user.getUserId(), tokenPair.refreshToken);
    await this.refreshTokenRepository.save(refreshToken);

    return {
      userId: user.getUserId(),
      username: user.getUsername(),
      email: user.getEmail(),
      accessToken: tokenPair.accessToken,
      refreshToken: tokenPair.refreshToken,
      tokenType: tokenPair.tokenType,
      expiresIn: tokenPair.expiresIn,
    };
  }

  async refresh(refreshTokenString: string): Promise<TokenPair> {
    // Hash the refresh token
    const tokenHash = RefreshToken.hashToken(refreshTokenString);

    // Find refresh token in database
    const refreshToken = await this.refreshTokenRepository.findByHash(tokenHash);
    if (!refreshToken) {
      throw AuthError.invalidToken();
    }

    // Validate token
    if (!refreshToken.isValid()) {
      if (refreshToken.isRevoked()) {
        throw AuthError.tokenRevoked();
      }
      throw AuthError.tokenExpired();
    }

    // Generate new token pair
    const tokenPair = this.jwtService.generatePair(refreshToken.getUserId());

    // Revoke old refresh token
    await this.refreshTokenRepository.revoke(refreshToken.getTokenId());

    // Save new refresh token
    const newTokenId = uuidv4();
    const newRefreshToken = RefreshToken.create(
      newTokenId,
      refreshToken.getUserId(),
      tokenPair.refreshToken
    );
    await this.refreshTokenRepository.save(newRefreshToken);

    return tokenPair;
  }

  async logout(refreshTokenString: string): Promise<void> {
    const tokenHash = RefreshToken.hashToken(refreshTokenString);
    const refreshToken = await this.refreshTokenRepository.findByHash(tokenHash);

    if (refreshToken) {
      await this.refreshTokenRepository.revoke(refreshToken.getTokenId());
    }
  }

  validateAccessToken(token: string): string {
    try {
      const claims = this.jwtService.verify(token);

      if (claims.tokenType !== 'access') {
        throw AuthError.invalidToken();
      }

      return claims.sub;
    } catch (error) {
      if (error instanceof Error) {
        if (error.message === 'Token expired') {
          throw AuthError.tokenExpired();
        }
        throw AuthError.invalidToken();
      }
      throw error;
    }
  }
}
